core/src/main/scala/kafka/server/instrumentation/ProduceRequestInstrumentation.scala

```diff
@@ -0,0 +1,193 @@
+   */
+  def toTimeTakenInEachStageMessage(instrumentation: ProduceRequestInstrumentation): String = {
+    val marks = instrumentation.marks
+
+    // Enforce Finish is marked, so when sliding there is at list a full window
+    marks.get(Stage.Finish) match {
+      case None => instrumentation.markStage(Stage.Finish)
+      case _ =>
+    }
+
+    marks
+      .toSeq                          // Iterate over stages
+      .sortBy(s => (s._2, s._1.id))   // in order of marked timestamp then defined order of stage (to prevent multiple stages finish within 1ms)
+      .sliding(2)                     // Make sliding window to diff time taken in each stage
+      .map { window =>
+        val iter = window.iterator
+        val (stage1, timestamp1) = iter.next()
+        val (_, timestamp2) = iter.next()
+        // Use the next stage's timestamp mark minus this stage's timestamp mark to get time spent on current stage
+        // Last stage is on finish, no diff to check, i.e. no need to log
+        s"${stage1.toString}: ${timestamp2 - timestamp1}"
+      }
+      .mkString("(", ", ", ")")
+  }
+
+  /**
+   * Based on the config (e.g. probability funnel), logs the detail of the instrumentation
+   *
+   * @param request This is required to obtain the total time for threshold funnel
+   * @param instrumentation The instrumentation to log
+   */
+  def maybeLog(request: RequestChannel.Request,
+               instrumentation: ProduceRequestInstrumentation): Unit = {
+    val totalTimeMs = nanosToMs(time.nanoseconds() - request.startTimeNanos)
+    if (totalTimeMs < config.thresholdToLogMs) {
+      logger.debug("totalTimeMs={} is below {}={}, skip logging",
+                   totalTimeMs,
+                   KafkaConfig.LiLongTailProduceRequestLogThresholdMsProp,
+                   config.thresholdToLogMs)
+      return
+    }
+
+    val coin = rnd.nextDouble()
+    if (coin > config.logRatio) {
+      logger.debug("Funneled out by random coin {} > configured ratio {}", coin, config.logRatio)
+      return
+    }
+
+    // Details like instrumentation.topicPartitionsInRequest cannot be obtained from request object,
+    // because the underlying details, at this moment, have been cleared by ProduceRequest#clearPartitionRecords()
+    logger.info(
+      s"totalTimeMs=${totalTimeMs}; acks=${instrumentation.requiredAcks}; "
+        + s"topic_partitions=${instrumentation.topicPartitionsInRequest.map(tp => s"${tp.topic()}-${tp.partition()}").mkString("(", ", ", ")")}; "
+        + s"instrumentationEachStageMs=${toTimeTakenInEachStageMessage(instrumentation)}; "
+        // Shrinking ISR info is important to the produce request.
+        // When there are dead followers, an acks=-1 produce may block until the leader discovered it is dead **AND** the full ISR shrinkage to complete.
+        + s"${classOf[PendingShrinkIsr].getSimpleName}_count=${replicaManager.numOfPartitionOfIsrState(classOf[PendingShrinkIsr])}"
```

Comment:
As a future work, it would be nice to have logs to help investigating why shrinking ISR is super slow in some cases when a broker is dead.

Reply:
Shrinking ISR is something responded by controller.  That's a control plane issue though.

The leader would take ~10 seconds to realize a broker is down, send shrink request, then wait for the control plane to propagate.

Reply:
Yeah, it is unclear why this process takes up to minutes. If we will further invest in this long latency investigation, this would worth exploring.